### DISCUSSION TITLE
React-Hook-Form not working with useImperativeHandle

### DISCUSSION THREAD

**Discussion by javadev-jef** [react-hook-form/react-hook-form#10303] [Q&A] [answered]
### Version Number

7.43.9

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-bug-with-useimperativehandle-ezpbrk

### Steps to reproduce

1. Visit codesandbox.
2. Check that the input component for `username` changes the value of `placeholder` when one of the select options is changed.
3. When we click on the submit button, all form fields are cleared, and the `onSubmit` method is not called (which makes no sense).
4. Also note that I'm using the `resetField` method when an option is selected in the select, but it doesn't work either.
5. Now when removing `useImperativeHandle` from Input and "updatePlaceHolder" from App, when we perform the above steps again, this problem does not occur.

### Expected behaviour

This looks like a bug to me, correct me if I'm wrong, please. But I expected the behavior of the form to remain the same.

### What browsers are you seeing the problem on?

Firefox, Chrome, Edge

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
This example looks complicated, let me know what you want to achieve and I can offer an alternative.

**Comment by leapful** [comment, marked as answer]
@javadev-jef 

Please read carefully React documentation on `useImperativeHandle` hook from https://react.dev/reference/react/useImperativeHandle#:~:text=Note%20that%20in%20the%20code%20above%2C%20the%20ref%20is%20no%20longer%20forwarded%20to%20the%20%3Cinput%3E

The `ref` object is altered via `useImperativeHandle` hook and no longer be a real HTML DOM element therefore react-hook-form cannot get value from input and validation error always occurs that blocks form submit.

Just use custom prop name like `inputRef` to get real HTML DOM element of the input.

Working demo: https://codesandbox.io/s/react-hook-form-bug-with-useimperativehandle-forked-48qrw8?file=/src/App.tsx

**Reply by bluebill1049** [reply]
thank you @leapful 🙏

**Comment by javadev-jef** [comment]
Thank you so much @bluebill1049 and @leapful. I saw the documentation for `useImperativeHandle` before opening this thread, but I can see that I didn't understand it very well. I thought that the `ref` was not forwarded to the input due to the context of the example in the documentation, but I see that I was wrong. @leapful, your example works great. Thank you again!
